src/C++/SSLSocketInitiator.cpp

```diff
@@ -318,83 +323,163 @@ void SSLSocketInitiator::doConnect( const SessionID& s, const Dictionary& d )
+            }
+            return SSL_HANDSHAKE_IN_PROGRESS;
+        }
+        else if (err == SSL_ERROR_SYSCALL)
+        {
+            getLog()->onEvent("SSL_connect failed with SSL error " + IntConvertor::convert(err) + ". Error stack:");
+
+            char errorBuffer[512];
+            unsigned long systemError;
+
+            while ((systemError = ERR_get_error()) != 0) {
+                ERR_error_string_n(systemError, errorBuffer, sizeof(errorBuffer));
+                getLog()->onEvent(errorBuffer);
+            }
+            getLog()->onEvent("End of error stack");
+
+            getLog()->onEvent(socket_get_last_error());
+
+            return SSL_HANDSHAKE_FAILED;
+        }
+        else {            
             getLog()->onEvent("SSL_connect failed with SSL error " + IntConvertor::convert(err));
-            return false;
+            return SSL_HANDSHAKE_FAILED;
         }
-        ERR_clear_error();
-        rc = SSL_connect(ssl);
     }
-    return true;
+
+    return SSL_HANDSHAKE_SUCCEDED;
 }
 
 void SSLSocketInitiator::onConnect( SocketConnector& connector, socket_handle s )
 {
-  SocketConnections::iterator i = m_pendingConnections.find( s );
-  if( i == m_pendingConnections.end() ) return;
+  getLog()->onEvent("Socket connected handle: " + std::to_string(s));
+   
+  time_t now;
+  ::time(&now);
+
+  SocketConnections::iterator i = m_pendingConnections.find(s);
+  if (i == m_pendingConnections.end()) return;
   SSLSocketConnection* pSocketConnection = i->second;
+  
+  m_pendingConnections.erase(i);
+  m_pendingSSLHandshakes[s] = pSocketConnection;
+  pSocketConnection->setHandshakeStartTime(now);
+    
+  handshakeSSLAndHandleConnection(connector, s); 
+}
 
-  bool sslHandshakeSuccess = handshakeSSL(pSocketConnection->sslObject());
+void SSLSocketInitiator::handshakeSSLAndHandleConnection(SocketConnector& connector, socket_handle s) {
+    SocketConnections::iterator i = m_pendingSSLHandshakes.find(s);
+    if (i == m_pendingSSLHandshakes.end()) return;
+    SSLSocketConnection* pSocketConnection = i->second;
 
-  if (sslHandshakeSuccess) {
-      m_connections[s] = pSocketConnection;
-      m_pendingConnections.erase(i);
-      setConnected(pSocketConnection->getSession()->getSessionID());
-      pSocketConnection->onTimeout();
-  }else
-  {
-      setDisconnected(pSocketConnection->getSession()->getSessionID());
-	  Session* pSession = pSocketConnection->getSession();
-      if (pSession)
-      {
-          pSession->disconnect();
-          setDisconnected(pSession->getSessionID());
-      }      
-      delete pSocketConnection;
-	  m_pendingConnections.erase(i);
-  }
+    SSLHandshakeStatus sslHandshakeStatus = handshakeSSL(pSocketConnection);
+
+    if (sslHandshakeStatus == SSL_HANDSHAKE_SUCCEDED) {
+        m_connections[s] = pSocketConnection;
+        m_pendingSSLHandshakes.erase(i);
+        setConnected(pSocketConnection->getSession()->getSessionID());
+        pSocketConnection->onTimeout();
```

Comment:
if handshake succeeds, removes the connection from the pendingSSLHandshakesList, and places it into the connections list.